File: src/Sigmoid.java
public class Sigmoid implements ActivationFunction {
    public Matrix apply(Matrix input) {
        Matrix output = new Matrix(input.rows, input.cols);
        for (int i = 0; i < input.rows; i++) {
            for (int j = 0; j < input.cols; j++) {
                output.data[i][j] = 1.0 / (1.0 + Math.exp(-input.data[i][j]));
            }
        }
        return output;
    }

    public Matrix derivative(Matrix input) {
        Matrix sig = this.apply(input);
        Matrix oneMinusSig = new Matrix(sig.rows, sig.cols);
        for (int i = 0; i < sig.rows; i++) {
            for (int j = 0; j < sig.cols; j++) {
                oneMinusSig.data[i][j] = 1.0 - sig.data[i][j];
            }
        }
        return sig.elementWiseMultiply(oneMinusSig);
    }
}

File: activation/Sigmoid.cpp
#include "Sigmoid.hpp"
#include <cmath>

Matrix Sigmoid::activate(const Matrix& input) const {
    Matrix result(input.rows, input.cols);
    
    for (int i = 0; i < input.rows * input.cols; ++i) {
        // Sigmoid: 1 / (1 + e^(-x))
        // Use stable computation to avoid overflow
        double x = input.data[i];
        if (x >= 0) {
            double exp_neg_x = std::exp(-x);
            result.data[i] = 1.0 / (1.0 + exp_neg_x);
        } else {
            double exp_x = std::exp(x);
            result.data[i] = exp_x / (1.0 + exp_x);
        }
    }
    
    return result;
}

Matrix Sigmoid::derivative(const Matrix& input) const {
    Matrix sigmoid_output = activate(input);
    Matrix result(input.rows, input.cols);
    
    for (int i = 0; i < input.rows * input.cols; ++i) {
        double s = sigmoid_output.data[i];
        result.data[i] = s * (1.0 - s);
    }
    
    return result;
}

std::string Sigmoid::getName() const {
    return "Sigmoid";
}
